flink-runtime/src/test/java/org/apache/flink/runtime/security/FlinkUserSecurityManagerTest.java

```diff
@@ -0,0 +1,187 @@
+
+	@Before
+	public void setUp() {
+		flinkUserSecurityManager = new FlinkUserSecurityManager(FlinkUserSecurityManager.CheckExitMode.THROW);
+		System.setSecurityManager(flinkUserSecurityManager);
+	}
+
+	@After
+	public void tearDown() {
+		if (flinkUserSecurityManager != null) {
+			System.setSecurityManager(flinkUserSecurityManager.getOriginalSecurityManager());
+		}
+	}
+
+	@Test(expected = UserSystemExitException.class)
+	public void testThrowUserExit() {
+		flinkUserSecurityManager.monitorSystemExit();
+		flinkUserSecurityManager.checkExit(TEST_EXIT_CODE);
+	}
+
+	@Test
+	public void testToggleUserExit() {
+		flinkUserSecurityManager.checkExit(TEST_EXIT_CODE);
+		flinkUserSecurityManager.monitorSystemExit();
+		try {
+			flinkUserSecurityManager.checkExit(TEST_EXIT_CODE);
+			fail();
+		} catch (UserSystemExitException ignored) { }
+		flinkUserSecurityManager.unmonitorSystemExit();
+		flinkUserSecurityManager.checkExit(TEST_EXIT_CODE);
+	}
+
+	@Test
+	public void testPerThreadThrowUserExit() throws Exception {
+		ExecutorService executorService = Executors.newSingleThreadExecutor();
+		// Async thread test before enabling monitoring ensures it does not throw while prestarting
+		// worker thread, which is to be unmonitored and tested after enabling monitoring enabled.
+		CompletableFuture<Void> future =
+			CompletableFuture.runAsync(() -> flinkUserSecurityManager.checkExit(TEST_EXIT_CODE),
+				executorService);
+		future.get();
+		flinkUserSecurityManager.monitorSystemExit();
+		try {
+			flinkUserSecurityManager.checkExit(TEST_EXIT_CODE);
+			fail();
+		} catch (UserSystemExitException ignored) { }
+		// This threaded exit should be allowed as thread is not spawned while monitor is enabled.
+		future = CompletableFuture.runAsync(() -> flinkUserSecurityManager.checkExit(TEST_EXIT_CODE),
+				executorService);
+		future.get();
+	}
+
+	@Test
+	public void testInheritedThrowUserExit() throws Exception {
+		flinkUserSecurityManager.monitorSystemExit();
+		try {
+			flinkUserSecurityManager.checkExit(TEST_EXIT_CODE);
+			fail();
+		} catch (UserSystemExitException ignored) { }
+		Thread thread = new Thread(() -> {
+			try {
+				flinkUserSecurityManager.checkExit(TEST_EXIT_CODE);
+				fail();
+			} catch (UserSystemExitException ignored) {
+			} catch (Throwable t) {
+				fail();
+			}
+		});
+		thread.start();
+		thread.join();
+	}
+
+	@Test
+	public void testWarnUserExit() {
+		// Warn mode enables monitor but only logging allowing exit, hence not expecting exception.
+		// NOTE - Do not specifically test warning logging.
+		System.setSecurityManager(flinkUserSecurityManager.getOriginalSecurityManager());
+		flinkUserSecurityManager = new FlinkUserSecurityManager(FlinkUserSecurityManager.CheckExitMode.WARN);
+		System.setSecurityManager(flinkUserSecurityManager);
+		flinkUserSecurityManager.monitorSystemExit();
+		flinkUserSecurityManager.checkExit(TEST_EXIT_CODE);
+	}
+
+	@Test
+	public void testValidConfiguration() {
```

Comment:
Could we split this up into multiple test cases/methods to improve readability?

Reply:
Will do.